app/components/admin/edition_images/uploaded_images_component.rb

```diff
@@ -26,11 +26,16 @@ def image_to_object(image, index)
       preview_alt_text: index.zero? ? "Lead image" : "Image #{index}",
       caption: image.caption.presence || "None",
       alt_text: image.alt_text.presence || "None",
-      markdown: "[#{image.image_data.carrierwave_image}]",
+      markdown: unique_names? ? "[#{image.filename}]" : "!!#{index}",
       links: links_for_image(image),
     }
   end
 
+  def unique_names?
+    names = @edition.images.map(&:filename)
+    names.uniq.length == names.length
+  end
```

Comment:
Let's [memoize](https://michaeljherold.com/articles/memoization-in-ruby/) this value to avoid an exponential loop. Otherwise we'll end up running `@edition.images.map` once for every image on the edition.

Since this is a boolean value, we can't use the usual `@unique_names ||=` approach to memoizing as that won't work for `false` values.

Instead, how about something like this:

```ruby
  def unique_names?
    return @unique_names if defined? @unique_names

    names = @edition.images.map(&:filename)
    @unique_names = (names.uniq.length == names.length)
  end
```

Or alternatively, a less 'clever' solution is just to run the method once in the class constructor, and then use `@unique_names` everywhere else:

```ruby
  def initialize(edition:)
    @edition = edition
    @unique_names = unique_names?
  end
```

Reply:
Good point, thanks Ollie, have updated the code